posthog/rate_limit.py

```diff
@@ -1,52 +1,134 @@
-from rest_framework.throttling import UserRateThrottle
+import re
+import time
+from functools import lru_cache
+from typing import List, Optional
+
+from rest_framework.throttling import SimpleRateThrottle
 from sentry_sdk.api import capture_exception
 from statshog.defaults.django import statsd
 
+from posthog.models.instance_setting import get_instance_setting
+from posthog.settings.utils import get_list
+
+
+@lru_cache(maxsize=1)
+def get_team_allow_list(_ttl: int) -> List[str]:
+    """
+    The "allow list" will change way less frequently than it will be called
+    _ttl is passed an infrequently changing value to ensure the cache is invalidated after some delay
+    """
+    return get_list(get_instance_setting("RATE_LIMITING_ALLOW_LIST_TEAMS"))
+
+
+@lru_cache(maxsize=1)
+def is_rate_limit_enabled(_ttl: int) -> bool:
+    """
+    The setting will change way less frequently than it will be called
+    _ttl is passed an infrequently changing value to ensure the cache is invalidated after some delay
+    """
+    return get_instance_setting("RATE_LIMIT_ENABLED")
+
+
+path_by_team_pattern = re.compile(r"/api/projects/(\d+)/")
+path_by_org_pattern = re.compile(r"/api/organizations/(.+)/")
+
+
+class PassThroughTeamRateThrottle(SimpleRateThrottle):
+    @staticmethod
+    def safely_get_team_id_from_view(view):
+        """
+        Gets the team_id from a view without throwing.
+
+        Not all views have a team_id (e.g. the /organization endpoints),
+        and accessing it when it does not exist throws a KeyError. Hence, this method.
+        """
+        try:
+            return getattr(view, "team_id", None)
+        except KeyError:
+            return None
 
-class PassThroughThrottle(UserRateThrottle):
-    # This class is being used as we're figuring out what our throttle limits should be.
-    # This ensures no rate limits are actually applied, but rather logs that they would have been applied.
-    # Allowing us to determine appropriate limits without affecting users.
     def allow_request(self, request, view):
+
+        if not is_rate_limit_enabled(round(time.time() / 60)):
```

Comment:
Maybe we should bust the cache more frequently for now, so we can turn this off quickly if it causes issues 🤔

Reply:
Last time this took down capture. That _shouldn't_ happen this time. But I guess 60 seconds is hugely faster than waiting for a revert and deploy so I'm probably over thinking it

e.g. is 20 seconds really better than 60 🤔 

That said I bet if we make it 20 seconds the outage would be because we're checking instance settings more frequently 🤣 So, I think let's deploy and see what happens

Reply:
checks out 😂